### PR TITLE
Reuse StorageController for LDS

### PR DESCRIPTION
**Pull request by dplewis**
Can be useful for https://github.com/parse-community/Parse-SDK-JS/issues/967

* Reduces the number of test and files.

Let me know if changes are needed.

Some of the promises could be written better.

### CODE REVIEW

File: src/LocalDatastoreController.js [hunk 1]
```diff
@@ -8,40 +8,40 @@
...
   },
 
-  getAllContents(): Object {
+  async getAllContents(): Object {
+    const keys = await Storage.getAllKeysAsync();
     const LDS = {};
-    for (let i = 0; i < localStorage.length; i += 1) {
-      const key = localStorage.key(i);
+
+    for(const key of keys){
```
Comment (acinader) [inline src/LocalDatastoreController.js]: this is fine, but this kind of pattern I would use reduce.

File: src/__tests__/LocalDatastore-test.js [hunk 2]
```diff
@@ -788,71 +789,14 @@ describe('LocalDatastore', () => {
...
-describe('LocalDatastore (BrowserDatastoreController)', () => {
-  beforeEach(async () => {
-    CoreManager.setLocalDatastoreController(BrowserDatastoreController);
+    CoreManager.setStorageController(BrowserStorageController);
+    CoreManager.setLocalDatastoreController(LocalDatastoreController);
     await LocalDatastore._clear();
   });
 
-  it('can store and retrieve values', async () => {
+  fit('can store and retrieve values', async () => {
```
Comment (acinader) [inline src/__tests__/LocalDatastore-test.js]: oops.

Comment (dplewis) [reply src/__tests__/LocalDatastore-test.js]: Just fixed that

File: src/LocalDatastoreController.js (hunk 1 above)
Comment (dplewis) [reply src/LocalDatastoreController.js]: Please help great one. I tried using reduce but gave up.

Comment (acinader) [reply src/LocalDatastoreController.js]: sorry, i'm wrong, localstorage is not iterable, so you can't use map and reduce on it...

Comment (dplewis) [reply src/LocalDatastoreController.js]: I'm looping through a set of keys and building a json object. localStorage isn't there anymore.

Comment (acinader) [reply src/LocalDatastoreController.js]: right, one sec.

Comment (acinader) [reply src/LocalDatastoreController.js]: https://runkit.com/arthur/reduce

Comment (dplewis) [reply src/LocalDatastoreController.js]: What about the async / await?

Comment (acinader) [reply src/LocalDatastoreController.js]: make the reducer function async:

```js
array.reduce(async(accumulator...
```

Comment (dplewis) [reply src/LocalDatastoreController.js]: That would be cool, I also figured out how to use reduce, I forgot to return

File: src/LocalDatastoreController.js [hunk 3]
```diff
@@ -23,49 +22,46 @@ const LocalDatastoreController = {
...
+    return Storage.removeItemAsync(name);
   },
 
-  getAllContents(): Object {
-    const LDS = {};
-    for (let i = 0; i < localStorage.length; i += 1) {
-      const key = localStorage.key(i);
+  async getAllContents(): Object {
+    const keys = await Storage.getAllKeysAsync();
+    return keys.reduce(async (previousPromise, key) => {
```
Comment (acinader) [inline src/LocalDatastoreController.js]: nice.

**Comment by Brokray** [comment]
Hello @dplewis,      

I just saw you added 'getAllKeys' as a required method for a `StorageController`, could you explain me why so? Is that intended?       

It seems to me that in the definition type of a StorageController there is no getAllKeys method expected.       
Which statement should we believe?     

[Here](https://github.com/parse-community/Parse-SDK-JS/blame/681016da6b521a0060c81e1efbe31ff0d07a3cc1/src/CoreManager.js#L367) you can find the change I am talking about.

**Comment by dplewis** [comment]
That is intended for dumping the LocalDatastore.

If the definition type is missing can you add it?

**Pull request merged**